A service worker should be able to focus a page that is not yet loaded

youennf opened the pull request:
#### 29c19d789902406975ea994e9838e33251b8422a
<pre>
A service worker should be able to focus a page that is not yet loaded
<a href="https://bugs.webkit.org/show_bug.cgi?id=279181">https://bugs.webkit.org/show_bug.cgi?id=279181</a>
<a href="https://rdar.apple.com/135337772">rdar://135337772</a>

Reviewed by Ben Nham.

Controlled service worker clients are exposed to service workers as soon as their navigation load hits the network process.
At this point, in the WebProcess, there is a DocumentLoader but not yet a Document.
We update WebSWClientConnection::focusServiceWorkerClient to get the WebPage from a DocumentLoader if there is no Document.

* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::pageFromScriptExecutionContextIdentifier):
(WebKit::WebSWClientConnection::focusServiceWorkerClient):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorker, FocusNotYetLoadedClient)):

Canonical link: <a href="https://commits.webkit.org/283263@main">https://commits.webkit.org/283263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f6054a50e7ff2758d4e2f84610a109f64136844

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65725 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69751 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16334 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16616 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52757 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11338 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68792 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41635 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56882 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33386 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38310 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14262 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15210 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60156 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14604 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71457 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9680 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14039 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60074 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56948 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60352 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14488 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7990 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1639 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40906 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41982 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43165 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
```diff
@@ -410,25 +410,35 @@ void WebSWClientConnection::notifyRecordResponseBodyEnd(RetrieveRecordResponseBo
-        RefPtr client = Document::allDocumentsMap().get(clientIdentifier);
-        RefPtr frame = client ? client->frame() : nullptr;
-        auto* page = frame ? frame->page() : nullptr;
+        RefPtr document = Document::allDocumentsMap().get(clientIdentifier);
```
I suppose there could still be an issue here if the DocumentLoader hasn't created the Document by the time this callback runs? Not sure what the consequences of that are.

youennf replied:
There is a compromise here, either we focus right away the tab even if we have not yet received the response.
Or we wait for the Document to proceed with the focus steps.
The spec is more inline with the latter.

youennf replied:
Maybe we should go to the UIProcess but wait for the Document creation to trigger these steps and resolve the promise.
I'll update the patch accordingly.

youennf replied:
Actually, I'll land it as is for now as it is an improvement.
I will work on waiting for the Document creation to resolve the focus promise in a follow-up PR.

youennf replied:
https://github.com/WebKit/WebKit/pull/33240

bnham reviewed (approved):
LGTM, we discussed how to make the API test work on iOS offline

youennf committed:
A service worker should be able to focus a page that is not yet loaded
https://bugs.webkit.org/show_bug.cgi?id=279181
rdar://135337772

Reviewed by Ben Nham.

Controlled service worker clients are exposed to service workers as soon as their navigation load hits the network process.
At this point, in the WebProcess, there is a DocumentLoader but not yet a Document.
We update WebSWClientConnection::focusServiceWorkerClient to get the WebPage from a DocumentLoader if there is no Document.

* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::pageFromScriptExecutionContextIdentifier):
(WebKit::WebSWClientConnection::focusServiceWorkerClient):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
((ServiceWorker, FocusNotYetLoadedClient)):

Canonical link: https://commits.webkit.org/283263@main
